source/material_model/equation_of_state/multicomponent_compressible.cc

```diff
@@ -0,0 +1,173 @@
+/*
+  Copyright (C) 2011 - 2019 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/equation_of_state/multicomponent_compressible.h>
+#include <aspect/utilities.h>
+
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace EquationOfState
+    {
+      template <int dim>
+      void
+      MulticomponentCompressible<dim>::
+      evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+               const unsigned int q,
+               MaterialModel::EquationOfStateOutputs<dim> &out) const
+      {
+
+        const double pressure = in.pressure[q];
+        const double temperature = std::max(in.temperature[q], 1.); // temperature can't be zero for correct evaluation
+
+        for (unsigned int c=0; c < out.densities.size(); ++c)
+          {
+            const double ak = reference_thermal_expansivities[c]/reference_isothermal_compressibilities[c];
+            const double f = (1. + (pressure - ak*(temperature - reference_temperatures[c])) *
+                              isothermal_bulk_modulus_pressure_derivatives[c] *
+                              reference_isothermal_compressibilities[c]);
+
+            out.densities[c] = reference_densities[c]*std::pow(f, 1./isothermal_bulk_modulus_pressure_derivatives[c]);
+            out.thermal_expansion_coefficients[c] = reference_thermal_expansivities[c] / f;
+            out.specific_heat_capacities[c] = (isochoric_specific_heats[c] +
+                                               (temperature*reference_thermal_expansivities[c] *
+                                                ak * std::pow(f, -1.-(1./isothermal_bulk_modulus_pressure_derivatives[c]))
+                                                / reference_densities[c]));
+            out.compressibilities[c] = reference_isothermal_compressibilities[c]/f;
+            out.entropy_derivative_pressure[c] = -out.thermal_expansion_coefficients[c] / out.densities[c];
+            out.entropy_derivative_temperature[c] = out.specific_heat_capacities[c] / temperature;
```

Comment:
We usually use these derivatives for latent heat generation at phase boundaries only, but the current implementation looks like it always produces derivatives. Let's have a chat about this, I feel like I mislead you, when you asked if these derivatives were isothermal/isobaric.

Reply:
Ok, corrected! Thanks for catching this.